packages/selected-network-controller/src/SelectedNetworkMiddleware.ts

```diff
@@ -0,0 +1,44 @@
+import type {
+  NetworkClientId,
+  NetworkControllerMessenger,
+} from '@metamask/network-controller';
+import type { JsonRpcMiddleware } from 'json-rpc-engine';
+
+import type { SelectedNetworkControllerMessenger } from './SelectedNetworkController';
+import { SelectedNetworkControllerActionTypes } from './SelectedNetworkController';
+
+const createSelectedNetworkMiddleware = (
+  selectedNetworkControllerMessenger: SelectedNetworkControllerMessenger,
+  networkControllerMessenger: NetworkControllerMessenger,
```

Comment:
Do we need to pass multiple messengers?

The goal of a messenger is to communicate with the entire system. That means there should be really only _one_ messenger that all controllers share.

Each controller does have a restricted messenger, but that's to restrict what messages controllers can send from inside themselves. However, all restricted messengers are backed by and proxy to a non-restricted messenger, so when you add an action handler through a restricted messenger it really gets added to the non-restricted one. That is, you can send messages from the non-restricted messenger even if they are registered via restricted messengers.

All this to say: I think we can just have this middleware take one messenger object, which is typed such that it recognizes three actions:

- `SelectedNetworkController:getNetworkClientIdForDomain`
- `SelectedNetworkController:setNetworkClientIdForDomain`
- `NetworkController:getState`

This should simplify the amount of setup you have to do for this test slightly.

Reply:
I had started out with the idea of having a single messenger passed in. I got bogged down in trying to make it a restricted messenger that you'd pass in, but I will try with a non-restricted messenger.